plugins/vpp/dnsplugin/descriptor/dnsserver.go

```diff
@@ -0,0 +1,149 @@
+		Update:          ctx.Update,
+	}
+	return adapter.NewDNSServerDescriptor(typedDescr)
+}
+
+// ValidateDNSServers validates content of DNS server configuration
+func (d *DNSServerDescriptor) ValidateDNSServers(key string, dnsServer *dns.DNSServer) error {
+	if len(dnsServer.UpstreamDnsServers) == 0 {
+		return scheduler.NewInvalidValueError(
+			errors.New("at least one upstream DNS server must be defined"), "upstreamDnsServers")
+	}
+	for _, serverIpAddress := range dnsServer.UpstreamDnsServers {
+		if net.ParseIP(serverIpAddress) == nil {
+			return scheduler.NewInvalidValueError(errors.Errorf("failed to parse upstream DNS Server IP "+
+				"address %s, should be a valid ipv4/ipv6 address", serverIpAddress), "upstreamDnsServers")
+		}
+	}
+	return nil
+}
+
+// EquivalentDNSServers determines whether 2 DNS servers are logically equal. This comparison takes
+// into consideration also semantics that couldn't be modeled into proto models (i.e. upstream servers are
+// IP addresses and not only strings)
+func (d *DNSServerDescriptor) EquivalentDNSServers(key string, oldDNSServer, newDNSServer *dns.DNSServer) bool {
+	if (oldDNSServer.UpstreamDnsServers == nil) != (newDNSServer.UpstreamDnsServers == nil) {
+		return false
+	}
+	if len(oldDNSServer.UpstreamDnsServers) != len(newDNSServer.UpstreamDnsServers) {
+		return false
+	}
+
+	for i := range newDNSServer.UpstreamDnsServers {
+		if newDNSServer.UpstreamDnsServers[i] != oldDNSServer.UpstreamDnsServers[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// Create enables and configures DNS functionality in VPP using VPP's binary api
+func (d *DNSServerDescriptor) Create(key string, value *dns.DNSServer) (metadata interface{}, err error) {
+	if err := d.updateUpstreamDNSServerList(nil, value.UpstreamDnsServers); err != nil {
+		return nil, err
+	}
+	if err := d.dnsHandler.EnableDNS(); err != nil {
+		return nil, errors.Errorf("failed to enable DNS due to: %v", err)
+	}
+	return nil, nil
+}
+
+// Delete disables (and removes configuration) DNS functionality in VPP using VPP's binary api
+func (d *DNSServerDescriptor) Delete(key string, value *dns.DNSServer, metadata interface{}) error {
+	if err := d.dnsHandler.DisableDNS(); err != nil {
+		return errors.Errorf("failed to disable DNS due to: %v", err)
+	}
+	if err := d.updateUpstreamDNSServerList(value.UpstreamDnsServers, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Update updates DNS Server configuration of already enabled DNS functionality
+func (d *DNSServerDescriptor) Update(key string, oldValue, newValue *dns.DNSServer,
+	oldMetadata interface{}) (newMetadata interface{}, err error) {
+	if err := d.updateUpstreamDNSServerList(oldValue.UpstreamDnsServers, newValue.UpstreamDnsServers); err != nil {
+		return nil, errors.Errorf("can't update upstream DNS servers due to: %v", err)
+	}
+	return nil, nil
+}
+
+func (d *DNSServerDescriptor) updateUpstreamDNSServerList(oldServers, newServers []string) error {
```

Comment:
Could be more efficient - delete only the obsolete entries and add only the truly new ones. Otherwise it is essentially re-create and so the Update method could be omitted.

Reply:
Initially i wanted to do it efficiently until i realized that it is not possible. VPP doesn't allow me to insert new upstream DNS server into proper place in the list, but just append at the end of list. Hence the recreate. However i forgot to reevaluate the usefulness of the Update function implementation.
Fixed in [`ac77bf2`](https://github.com/ligato/vpp-agent/pull/1779/commits/ac77bf2faa7c4b796216c7917a545284c73d97d3).